Compile variable features

simoncozens opened the pull request:
This is a work in progress. The idea is this: when we are compiling a variable font from designspace, most of the time the feature files are exactly the same; the only thing which differs is the "implicit" rules created from anchors and kerning. Currently we build each font independently, and this means we compile binary features N times (where N=number of masters), and then merge them at the end. When there are overflows, compiling binary features N times is slow. Instead, this PR holds off the compilation of GSUB/GPOS tables to the end, and uses "variable" feature writers which describe how the kerning and anchors vary across the designspace. i.e. instead of:

```fea
### MutatorMathTest-LightCondensed ###
lookup kern_ltr {
    lookupflag IgnoreMarks;
    pos A J -10;
} kern_ltr;

### MutatorMathTest-BoldCondensed ###
lookup kern_ltr {
    lookupflag IgnoreMarks;
    pos A J -20;
} kern_ltr;
```

and compiles the features twice, it produces

```fea
lookup kern_ltr {
    lookupflag IgnoreMarks;
    pos A J (wght=0:-10 wght=1000:-20);
} kern_ltr;
```

and compiles the features once.

It's a lot faster.

- [x] Curs feature writer is untested.
- [x] There's some problem with enumerated kerns.
- [x] I'm not sure I trust it yet.

madig commented on Lib/ufo2ft/__init__.py:
```diff
@@ -738,6 +753,18 @@ def compileVariableCFF2s(designSpaceDoc, **kwargs):
+        return text
+
+    first = transform(designSpaceDoc.sources[0])
+    return all(transform(s) == first for s in designSpaceDoc.sources[1:])
```
At this stage, features are not expanded to masters, so differing imports are not checked. Is this a problem?

simoncozens replied:
Not sure I understand this. What does expanding features to masters mean?

madig replied:
Resolving all imports with `parseLayoutFeatures`. We can't test for equality what we can't get our hands on. I suppose that this is an edge case, if you use feature file syntax without variation conditionals, checking just the master feature text should be enough, because you can't change master behavior elsewhere without it being imported everywhere else.

simoncozens replied:
If the files are identical, any imports will import the same files. So that case is OK. If the UFOs import different feature files which turn out to be the same code, then the font engineer is pathological and deserves to pay a small performance penalty.

madig replied:
Never discount the possibility of a small bug somewhere going unnoticed :)

madig commented:
So to trigger this, a font project must make sure to have the same exact feature file text in every source, meaning that you have to use variable feature syntax if there are differences in some parts of the Designspace?

madig commented on Lib/ufo2ft/featureCompiler.py:
```diff
@@ -92,6 +95,10 @@ def __init__(self, ufo, ttFont=None, glyphSet=None, **kwargs):
         glyphOrder = ttFont.getGlyphOrder()
         if glyphSet is not None:
+            if set(glyphOrder) != set(glyphSet.keys()):
+                print("Glyph order incompatible")
```
This should probably be a `logger.error`.

anthrotype replied:
this is just before an `assert` so it doesn't matter much if we `print` or do `logger.error`, an AssertionError should never trigger and is not something client can expect to recover from. It's just the error message for the hard crash, it could be moved to the `assert` optional string parameter, if you want.

simoncozens commented:
Yes, it doesn’t attempt to do a text/AST-based merge of the feature files. At least not at this stage - maybe that can come later. I’m fairly sure that this will cover a good number of font projects, and those that it doesn’t get cover will get the old style merge.

madig commented on Lib/ufo2ft/featureWriters/baseFeatureWriter.py:
```diff
@@ -306,19 +307,24 @@ def compileGSUB(self):
             glyphOrder = compiler.ttFont.getGlyphOrder()
+            fvar = compiler.ttFont.get("fvar")
+            if fvar:
+                feafile = copy.deepcopy(feafile)
```
Why the deepcopy?

anthrotype replied:
indeed, why?

madig commented on Lib/ufo2ft/featureWriters/variableRulesWriter.py:
```diff
@@ -0,0 +1,51 @@
+from ufo2ft.featureWriters import BaseFeatureWriter
+
+
+class VariableRulesFeatureWriter(BaseFeatureWriter):
```
This looks unused?

madig commented:
In one of our fonts, this seems to drop some kerning but not others 🤔 weird, drilling some more.

Also, the Rules writer is not hooked up and justr dropping it into the default feature writer list doesn't seem to do anything.

madig commented on Lib/ufo2ft/featureWriters/variableRulesWriter.py:
```diff
@@ -0,0 +1,51 @@
+                    self._conditionsets.append(conditionset)
+                else:
+                    cs_name = "ConditionSet%i" % self._conditionsets.index(conditionset)
+                # XXX
```
?

madig commented:
So, I get rules to work when I graft the GSUB `<FeatureVariations>` from a previous ufo2ft and adapt the lookup numbers. This means that the rules writer should do whatever varLib is doing I guess.

madig commented:
Weird. For my test font, no GSUB variations are emitted without the rules writer.

madig commented:
Now we do approximately what varLib is doing with rules, though compared to a reference, the numbers and lookups are off in some places. Maybe I use `overlayFeatureVariations` wrong. At least with inspection by eye, stuff seems to work as expected.

There still remains a difference in kern behavior I need to investigate.

madig commented:
Also todo: The test should include disjoint kerning pairs for different scripts, to ensure we handle disjoint kerning in different masters correctly (i.e. Armenian only in one, Greek only in the other).

madig commented:
> I disabled the rules writer because varLib also handles designspace rules during the merge and I was ending up with them being emitted twice.

Oh, I know what's going on. My test sources happen to trigger the compile-features-once-but-later code path, so varLib adds feature variations and then we overwrite them when running the feature writers. This also seems to affect kerning.

madig commented:
Next, my test source has e.g. an `r` vs `d` class-based kern that gets lost when compiling features later for VFs. 🤔

simoncozens commented:
You're using non-public test sources, which makes it difficult for other people to repro... I'll see what I can find.

simoncozens commented:
Note to self: `_featuresCompatible` will return false if there are any ligature caret positions, which glyphsLib writes into the feature file (with per-master-file variations). Need to think how to handle these.

madig commented:
Woops, yeah. I'll be back at work next week and can see if I can make a reproducer. For now, maybe just compile the font with identical features with 2.28.0 and then with this branch and compare kerning.

I thought I turned ligature writing off? Should be in https://github.com/googlefonts/glyphsLib/releases/tag/v6.0.7

anthrotype commented:
yeah, the GdefFeatureWriter in ufo2ft should handle both the GlyphClassDefs (using public.openTypeCategories in the ufo lib) and the LigatureCarets (from the glyph anchors named "caret_", "vcaret_")

madig commented:
So, I tried to come up with a minimal reproducer, but so far, no luck. What did seem to work was to make a union of all kerning values over all sources, putting in missing nones as zero-kerns. The specific pair I was looking at was actually "missing" such a zero-kern in one corner of  the design space. Maybe that's what's missing?

madig commented:
varLib allows "sparse kerning", i.e. if Light and bold have a value but Regular doesn't, the value at Regular is interpolated rather than set to zero. So, filling in zero for missing kern pairs seems like the wrong thing to do.

madig commented:
In my testing with a proprietary font, it seems the kerning is there, but it is not used, as if the shaper was stopping earlier. Subsetting the font to contain just the two glyphs I want to look at makes the kern work as expected. Will try to reproduce with an OSS font.

simoncozens commented:
Here is a very minimal reproducer:
[RobotoSerif-Mini.glyphs.gz](https://github.com/googlefonts/ufo2ft/files/9417330/RobotoSerif-Mini.glyphs.gz)

simoncozens commented:
With your patch and https://github.com/fonttools/fonttools/pull/2776, I think this is good now.

It has a slight change of behaviour compared to the past in respect of "empty" kern pairs, but I think this is for the better. If an intermediate master previously did not define a kern value for a pair which had kerning in other masters, we would assume that value was zero. For example, if you had Regular=-100, Bold=-200 and no kern defined in the SemiBold, you would get a kerning value of zero. Presumably designers would have realised that they didn't mean that, and put an explicit kern in there.

But now we can do better than that: if there is no explicit kern for an intermediate master, it will be interpolated. This should make people's kerning processes a lot faster...

simoncozens commented:
The remaining test failures are due to the new behaviour described above.

simoncozens commented:
OK, so one more bug we found: when backfilling kerns for missing masters, we shouldn't just use zero. varLib.merge also looks at more *general* rules in other masters and backfills using their value. e.g. if you have

```
enum pos @kern.1.o w -200;
pos @kern1.o @kern2.w -30;
```

in the regular master and 

```
pos @kern1.o @kern2.w -20;
```

in the bold, you *don't* want to do this:

```
enum pos @kern.1.o w (wght=400:-200 wght=800:0); # No equivalent rule, use zero
pos @kern1.o @kern2.w (wght=400:-30 wght=800:-20);
```

because that ends up stealing the kern altogether. instead you want to do this:

```
enum pos @kern.1.o w (wght=100:-200 wght=400:-20); # A more general rule covers this pair, use it
pos @kern1.o @kern2.w (wght=400:-30 wght=800:-20);
```

madig commented:
Talking to Jany, Harry and Cosimo, we need to figure out all cases of this and what should happen in each. 🤔 Maybe it is even not possible to keep the split-kerns-by-script code and make it work with the existing merge flow? We'd need to figure out what to do if one mastr does not have kerning for something, just like here, because otehrwise varLib complains about https://github.com/googlefonts/ufo2ft/issues/642.

simoncozens commented:
I don’t think there’s a mystery about what needs to happen.

If there is no exact equivalent kern pair in a master, the code should look to see if there is a kern provided by another class based rule, and if so backfill using that. If not it should use zero. This is exactly what varLib does in https://github.com/fonttools/fonttools/blob/main/Lib/fontTools/varLib/merger.py#L347-L355

madig commented:
I think I am beginning to understand. This WIP commit seems to make basic text mostly match ufo2ft 2.28.0-compiled Roboto Serif, expcet for the "ft" pair, which seems to have a positive kern in the 2.28.0 version.

madig commented:
Actually, can we reuse `fontTools.ufoLib.kerning.lookupKerningValue`? I see some other kerning diffs in Roboto Serif like in "f," and "øj", though.

madig commented:
With the latest commits, I can only see kerning diffs for the following two pairs on my testing page:

* øj
* øy

madig commented:
One thing to do would be to think about sparse kerning ([SparseKern.zip](https://github.com/googlefonts/ufo2ft/files/9464095/SparseKern.zip)) -- The Regular does not have a kernig value for AB. varLib will ensure there is no kerning value for it and the value will be interpolated, the latest commits here will insert a zero, making the kerning interpolate to zero in Regular. Sparse kerning should be supported.

simoncozens commented:
I see a 20 unit kern for øj  and øy with this PR. However, it looks like we have managed to make the build considerably slower. A large number of `Already defined position for pair...` messages too.

madig commented:
I noticed that the code doesn't seem to take into consideration sparse layers (Designspace sources with a layerName). Maybe it should.

madig commented:
Rebased on main, now to fix the tests... currently I don't know what's going on with them.

madig commented:
Note to self: if the sources are kerned such that we have e.g. `pos @kern1.Cyrl.PUNCT_hyphen @kern2.Cyrl.CYR_De_UC_2ND (wght=100:-20 wght=900:20);`, is it reasonable to interpolate the value at `wght=400` to appease the variation model instead of inserting a zero? I believe so. Still, find out what varLib does.

madig commented:
Looking at a flood of warnings about regrouped glyphs, but the code makes me think they're bogus... Still trying to figure them out.

madig commented:
FWIW, this produces kerning different from the merge path. I think I have to work on kerning-validator to 1. support comparing kerning between two fonts and 2. replicate the behaviour of varlib in determining the kerning value of a pair in a design space.

madig commented:
I'm beginning to suspect that there is something else going on during compilation. I have this big font here that has seemingly identical kerning when I compile the wght-only variant, but has "missing kerning" when I compile the bigger wght+wdth+ital variant. The kern I looked at had an entry in the debug feature file, but the wght+wdth+ital TTF showed no kerning for it. Need to find an open font I can use for repros.

madig commented:
I think I figured out the problem, at least for Roboto Serif. Compiling it shows missing kerning for the text "g,", but only at some locations and not others.

I subset the Italic Designspace (the uprights don't compile) to have just `space`, `g` and `comma` (using https://github.com/googlefonts/gftools/blob/main/Lib/gftools/ufomerge.py to merge it into an empty font, i.e. subset it this way) and then compiled a VF. The debug feature file showed this (edited for brevity, `GRAD=0,opsz=20,wdth=100,wght=400` is the default location, the other one some extreme end):

```fea
lookup kern_Latn {
    lookupflag IgnoreMarks;
    enum pos @kern1.Latn.g.alt comma (
        # ...
        GRAD=0,opsz=20,wdth=100,wght=400:0
        # ...
        GRAD=100,opsz=8,wdth=150,wght=900:-30
    );
    pos @kern1.Latn.g.alt @kern2.Latn.period (
        # ...
        GRAD=0,opsz=20,wdth=100,wght=400:-20
        # ...
        GRAD=100,opsz=8,wdth=150,wght=900:-40
        # ...
    );
} kern_Latn;
```

In the sources, the default location has a kerning pair 
```
pos @kern1.Latn.g.alt @kern2.Latn.period -20
```
and the extreme end at `GRAD=100,opsz=8,wdth=150,wght=900` has the kerning 
```
enum pos @kern1.Latn.g.alt comma -30
pos @kern1.Latn.g.alt @kern2.Latn.period -40
```

Since the kerning exception is only present in a few sources, but not the default one, the current code inserts a zero. When shaping "g,", the shaper grabs the first pair that matches, sees a zero, and stops. If it looked at the default location source in isolation, it would not find the exception and go on to the class to class kerning, resulting in the expected -20.

So... what to do about it. I think we need to handle exceptions specially by backfilling not a zero or nothing if it does not exist in all locations but the value from "higher up". In this case, the class-to-glyph pair would carry all values of the class-to-class value but overwrite some with its own?

I think the lookupKerningValue from ufoLib would do that, but I used it before and it had the same problem, so maybe I held it wrong.

Maybe I could do an extra pass after having collected all kerning from all sources, to find kerning exceptions and backfill missing locations (i.e. all class-to-glyph and glyph-to-class kernings would inherit values from the class-to-class parent, then all glyph-to-glyph pairs would inherit from their class-to-glyph and glyph-to-class parents) 🤔

madig commented:
I believe this script does what needs to be done: https://gist.github.com/madig/76567a9650de639bbff51ce010783790, both for variable feature compilation and the old merge-to-VF flow, _and_ instance interpolation, unless I'm missing something @Hoolean. If so, I'd need to work this into kerning-validator as a model.

Hoolean commented:
> I believe this script does what needs to be done: https://gist.github.com/madig/76567a9650de639bbff51ce010783790, both for variable feature compilation and the old merge-to-VF flow, _and_ instance interpolation, unless I'm missing something @Hoolean. If so, I'd need to work this into kerning-validator as a model.

The script's union kerning step largely maintains fontmake's variation model but we need to cover an additional case:

- ✅ For **sources with kerning and/or GPOS features**, the current behaviour is correct
- :x: For **other sources**, the current behaviour will insert 0-value kerns at every pair, where the values should be interpolated at this source's position in the final font instead

If our goal is to make every kerning value explicit at the source-level, the latter case may be difficult to handle without re-implementing feature interpolation.

Hoolean commented:
My above comment can be largely discounted after we revisited googlefonts/fontmake#998. Having discussed with @anthrotype, **we only need to maintain interpolated kerning for sparse layers**, and can drop the edge case of interpolated kerning for UFO sources that have an empty or absent kerning.plist and GPOS features:

* The behaviour likely originates from a bug with how we invoke varLib;
* The scenario is highly-specific and awkward to achieve (most sources have at least some GPOS features, e.g. even if implicitly from anchors);
* Automatic testing of a corpus of fonts found no sources using this behaviour; and
* Interpolation can be easily replicated by inserting explicit values, which is used in all other scenarios.

Sparse layers only contribute glyph outlines, thus neither contribute kerning nor participate in kerning interpolation; any kerning at their location will be interpolated at runtime from other kerning-contributing sources. All other sources have a kerning dictionary associated with them, and should participate in the kerning interpolation even if their kerning.plist is absent or empty; they should implicitly contribute zeros for the union of kerning pairs across sources.

This simplifies what we expect for kerning in the final font to:

* Sparse layers: fully interpolated
* All other layers: no interpolation

m4rc1e commented:
I've just hit the dreaded https://github.com/googlefonts/fontmake/issues/894 issue with Merriweather VF so I don't mind getting through @Hoolean's outstanding tasks. This solution does seem better than the previous attempts so I don't mind investing some time into this.

Hoolean commented:
@m4rc1e Hi Marc, happy to chat in more depth if this would be helpful for that :)

This PR – when completed – should indeed be able to avoid that fontMake issue, although it would depend on the feature files being identical across Merriweather's sources, probably relying on using variable syntax in some places :+1:

simoncozens commented:
I actually need this functionality for a project now, so I've rebased this PR. However, I still see a few problems:

* Mark anchors positioning (`_top` etc.) is not varied. (I'll fix this.)
* Something is wrong with cursive positioning. (I'll also try to fix this.)
* Integration tests are failing and giving significantly fewer variation regions for me locally, although I notice they're succeeding on the CI?!

anthrotype commented:
maybe you and CI are using different fonttoolses?

anthrotype commented:
CI is pinned at https://github.com/googlefonts/ufo2ft/blob/efb4658890ba645292a9c6bc055c91afe225c4a1/requirements.txt#L1

simoncozens commented:
OK, those issues are fixed, but I found a new one: we've lost the public.openTypeCategories information and now marks are missing from my GDEF table.

anthrotype commented:
> we've lost the public.openTypeCategories information

is that stored at the single-master level? we should probably try fetch it from the base master or preferably from the DS.lib because that should really be a global kind of data that applies to all masters

simoncozens commented:
Yeah, we can try both. OK, this is working for my purposes now. I'm not sure what the state of the kern debate is.

simoncozens commented:
Something weird happening with (ligature) mark positioning, although the numbers look right for me in the variable feature file...

madig commented:
> I'm not sure what the state of the kern debate is.

What debate?

simoncozens commented:
The ligatures problem is a fontTools ones, PR coming soon.

anthrotype commented:
> The ligatures problem is a fontTools ones, PR coming soon.

@simoncozens https://github.com/fonttools/fonttools/releases/tag/4.44.1

madig commented:
> but I am not sure what's the status of the "next steps" that he mentioned in https://github.com/googlefonts/ufo2ft/pull/635#issuecomment-1589588070

Probably untouched.

Hoolean commented:
> but I am not sure what's the status of the "next steps" that he mentioned in https://github.com/googlefonts/ufo2ft/pull/635#issuecomment-1589588070

Hello all! Yes, can confirm that the state of the Next Steps above is accurate; we made the main changes to the kerning but it still needs more thorough testing.

We began to adapt [kerning-validator](https://github.com/daltonmaag/kerning-validator) to test VFs but I have not had availability to complete this since, and so if anyone wishes to resume this ufo2ft PR then there would be no duplicated effort (and it would be much appreciated too :rocket:).

simoncozens commented:
This works beautifully on variable builds but seems to have broken static builds.

anthrotype commented:
broken how? Aren't these only used for variable builds?

simoncozens commented:
I'm still trying to track it down, but I'm seeing

```
  File "/Users/simon/others-repos/ufo2ft/Lib/ufo2ft/featureWriters/baseFeatureWriter.py", line 139, in write
    self.setContext(font, feaFile, compiler=compiler)
  File "/Users/simon/others-repos/ufo2ft/Lib/ufo2ft/featureWriters/variableMarkWriter.py", line 20, in setContext
    font=self.context.font.findDefault().font,
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Font' object has no attribute 'findDefault'
```

during a static build.

simoncozens commented:
Oh, I bet it's my glyphsLib thing, actually (https://github.com/googlefonts/glyphsLib/pull/957). GlyphsLib's mark feature writer needs to be variable or not dependent on the build.

anthrotype commented:
> I should re-organise this such that the feature writers do the right thing based on whether they're handed a font or a designspace

yeah, that may be preferable indeed

anthrotype commented:
have you rebased this on main? I think I have recently bumped fonttools and updated the integration tests in https://github.com/googlefonts/ufo2ft/pull/795

newer fonttools correctly prunes unused VarRegions (like in the above ttx dump) when optimizing VarStore

anthrotype commented:
or if you can't rebase, just git fetch origin && git merge origin/main

simoncozens commented:
Yes, this was rebased yesterday and includes f51c9f7f. The problem was that the branch also included another commit which also updated the expectations! So I've dropped that and now I have this change:

```
           <VarRegionAxis index="0">
-            <StartCoord value="0.0"/>
+            <StartCoord value="0.36365"/>
             <PeakCoord value="1.0"/>
             <EndCoord value="1.0"/>
           </VarRegionAxis>
```

This is with a designspace which has an intermediate layer at wght=450, and a mark positioning rule like so:

```
                    pos base e
                        <anchor (wght=350:314 wght=450:314 wght=625:315) (wght=350:556 wght=450:556 wght=625:644)> mark @MC_top;
```

I'm going to argue that this change (starting the variation at 450, not 350) is actually correct and that the current behaviour of ufo2ft is wrong: it doesn't respect variations with respect to marks in intermediate layers.

anthrotype commented:
> the current behaviour of ufo2ft is wrong: it doesn't respect variations with respect to marks in intermediate layers.

that's right, I think this is => https://github.com/googlefonts/ufo2ft/issues/640

simoncozens commented:
It's fairly gross that we have an integration test for behaviour that we know to be wrong. ;-) Anyway, fixed now.

anthrotype commented on Lib/ufo2ft/__init__.py:
```diff
@@ -798,6 +823,28 @@ def compileVariableCFF2s(designSpaceDoc, **kwargs):
+    """Returns whether the features of the individual source UFOs are the same.
+
+    NOTE: Only compares the feature file text inside the source UFO and does not
+    follow imports. This will suffice as long as no external feature file is
+    using variable syntax and all sources are stored n the same parent folder
+    (so the same includes point to the same files).
```
why can't external  feature file use variable syntax for features to be compatible?

```suggestion
    follow imports. This will suffice as long as all sources are stored n the same
    parent folder (so the same includes point to the same files).
```

simoncozens replied:
@madig wrote that bit, any thoughts? I think it should be fine to use variable syntax in external feature files.

madig replied:
I may have mistyped. What I meant is that all is good if files are identical and none are using some ifdef kind of text shenanigans.

anthrotype commented on tests/data/TestVarfea (Autosaved).glyphs:
```diff
@@ -0,0 +1,653 @@
+{
```
@simoncozens  this (Autosaved) file was probably saved by mistake

anthrotype commented on Lib/ufo2ft/__init__.py:
```diff
@@ -872,6 +928,62 @@ def _compileNeededSources(
+            featureCompiler.writeFeatures(debugFeatureFile)
+
+    # Add back feature variations, as the code above would overwrite them.
+    designSpaceData = varLib.load_designspace(designSpaceDoc)
+    featureTag = designSpaceData.lib.get(
+        varLib.FEAVAR_FEATURETAG_LIB_KEY,
+        "rclt" if designSpaceData.rulesProcessingLast else "rvrn",
+    )
+    if designSpaceData.rules:
+        varLib._add_GSUB_feature_variations(
+            ttFont,
+            designSpaceData.axes,
+            designSpaceData.internal_axis_supports,
+            designSpaceData.rules,
+            featureTag,
+        )
```
maybe we should add a new public method to varLib that does just that, instead of using a private method like this. Something that takes a DS and a VF, and translates all the DS rules into equivalent featureVars

anthrotype replied:
we should now be able to use the varLib.addGSUBFeatureVariations function that I added with fonttools 4.46: https://github.com/fonttools/fonttools/pull/3362

anthrotype commented on Lib/ufo2ft/__init__.py:
```diff
@@ -872,6 +928,62 @@ def _compileNeededSources(
+
+    # Add anything we added to the TTF without telling the UFO
+    if ".notdef" not in glyphSet:
+        glyphSet[".notdef"] = StubGlyph(".notdef", 0, 0, 0, 0)
```
I don't like that we have to do this here.. Ideally we would pass down the same glyphSet that was preprocessed and potentially expanded by the outlineCompiler

simoncozens replied:
Ideally, yes. But that would require having the glyphsets handy in `compileVariableTTFs`, which would mean returning them from `_compileNeededSources`, which would mean returning them from `compileInterpolatableTTFsFromDS`, which would mean returning them from `compileInterpolatableTTFs`. Major surgery.

anthrotype replied:
yes, it can't be otherwise

anthrotype replied:
the default master's glyphSet is by definition a superset of the glyphsets of the non default sources

simoncozens commented on Lib/ufo2ft/__init__.py:
```diff
@@ -659,6 +666,10 @@ def compileVariableTTFs(designSpaceDoc: DesignSpaceDocument, **kwargs):
     logger.info("Building variable TTF fonts: %s", ", ".join(vfNameToBaseUfo))
 
+    if featuresNeeded:
+        # Skip generating feature variations in varLib.
```
I've fixed this, but it's shown up that there's considerable code repetition between e.g. `compileVariableCFF2s` and `compileVariableTTFs`. I'm starting to think that these "compile a thing" functions may actually be better as *classes* rather than as *functions*. That would also help with the glyphsets plumbing problem (see comment below). So I might look into doing that major surgery.

madig commented:
Has the kerning getter been updated to do what https://gist.github.com/madig/76567a9650de639bbff51ce010783790#file-align-groups-py-L122-L154 is doing? I can't tell.

anthrotype commented:
> Has the kerning getter been updated to do what https://gist.github.com/madig/76567a9650de639bbff51ce010783790#file-align-groups-py-L122-L154 is doing? I can't tell.

@madig I think so, I can see the `lookupKerningValue` being used in the kernFeatureWriter, I think @Hoolean added that in https://github.com/googlefonts/ufo2ft/commit/effd4baeb1c5a1897079b357e82b291455c8ca4f

simoncozens commented:
Stand by for all kinds of git madness to get this vaguely rebaseable.

anthrotype commented on Lib/ufo2ft/util.py:
```diff
@@ -243,13 +243,18 @@ def makeUnicodeToGlyphNameMapping(font, glyphOrder=None):
     font = ttLib.TTFont()
     font.setGlyphOrder(glyphOrder)
-    addOpenTypeFeatures(font, featureFile, tables={"GSUB"})
+    if fvar:
+        font["fvar"] = fvar
+        tables = {"GSUB", "GDEF"}
```
can you add a comment why you're also building GDEF here when fvar is not None?

simoncozens replied:
Not needed.

simoncozens commented:
Right, I'm done here. Remember that without https://github.com/googlefonts/glyphsLib/pull/957 you will get wrong answers compiling Glyphs files.

anthrotype commented:
Thanks Simon, I really appreciate this effort.
So, I tried to run this on ramdom font, Oswald, and compared the ttx outputs for the fonts built with ufo2ft main vs compile-variable-features branch and I see a lot of diffs in GDEF/GPOS.
I am a bit hesitant to enable this by default and was wondering if we could add this feature behind an (opt-in) option?
We could also detect if any of the hand-written FEA contains any variable FEA syntax and if so compile them using the new approach, otherwise keep using the old one unless the user requests otherwise?

anthrotype commented:
regarding some of the diffs, I notice for example that the GDEF ligature carets are no longer variable (Format=1 instead of Format=3 CaretValue with accompanying delta set indices) for the font compiled using `compile-variable-features` branch.
Is this a known limitation?

anthrotype commented:
note to self, the new https://github.com/googlefonts/fontc/tree/main/layout-normalizer (used by fontc to compare vs fontmake) could also be used to compare the GPOS tables from varlib-merged font vs variable FEA font

simoncozens commented:
> regarding some of the diffs, I notice for example that the GDEF ligature carets are no longer variable (Format=1 instead of Format=3 CaretValue with accompanying delta set indices) for the font compiled using `compile-variable-features` branch. Is this a known limitation?

This wasn't known (at least not to me!) and, yes, it should be fixed.

> note to self, the new https://github.com/googlefonts/fontc/tree/main/layout-normalizer (used by fontc to ttx_diff vs fontmake) could also be used to compare the GPOS tables from varlib-merged font vs variable FEA font

I'd appreciate looking at these normalised diffs.

anthrotype commented:
you can run the fontc' layout-normalizer like this, from the fontc repository's root:

```
$ cargo run -p layout-normalizer -t GPOS -o font1.txt font1.ttf
$ cargo run -p layout-normalizer -t GPOS -o font2.txt font2.ttf
```

then you can diff the two text files

I can already say that for Oswald at least, there are _no_ differences in the kern feature, which I think counts as a great accomplishment! Well done us.

The diffs are only in the mark/mkmk and don't look too big. You can take a look at the files below, or just regenerate as above:

[Oswald-layout-normalizer-output.zip](https://github.com/googlefonts/ufo2ft/files/13582478/Oswald-layout-normalizer-output.zip)

anthrotype commented:
> Those diffs are just mark quantisation

did that change with this PR?

simoncozens commented:
OK, it's not quantisation but rounding. In the sources of Oswald's bold master, the brevebelowcomb has a bottom anchor at Y=-194.525. Current ufo2ft otRounds that to -195; this PR does not otRound a variable scalar, so it ends up as being -194. I think the fix is probably to otRound the number when it's being added to the variable scalar.

anthrotype commented on Lib/ufo2ft/_compilers/baseCompiler.py:
```diff
@@ -294,13 +323,60 @@ def compile_variable(self, designSpaceDoc):
             vfNameToTTFont = self._merge(designSpaceDoc, excludeVariationTables)
 
+        if buildVariableFeatures:
+            self.compile_all_variable_features(
```
can we log.info here that we are finally compiling all the variable features?

anthrotype commented:
Ok so @cmyr fixed a few bugs in the layout-normalizer so I have tried to run it again on the Oswalds generated with fontmake-main vs with this PR applied, and I now see there are a lot of missing kerning pairs in the font built with ufo2ft/compile-variable-features.
I attache the output from layout-normalized below. (I have used this branch https://github.com/googlefonts/fontc/pull/654 as it includes all bugfixes plus adds support for extension lookups, not merged to main yet).

@simoncozens can you please take a look?

[Oswald-layout-normalizer-output-2.zip](https://github.com/googlefonts/ufo2ft/files/13673282/Oswald-layout-normalizer-output-2.zip)

anthrotype commented:
actually, it's also possible that the layout-normalizer has some other bugs, because IIRC it was previously reporting the kerning to be the same for Oswald at least, and this PR didn't change in the meantime (the normalizer did). Sorry for the noise..

simoncozens commented:
Yep, I found the same, and then when I actually *tried* the kern pairs at the reported locations I got identical results between the two files. So it's worth checking manually.

anthrotype commented:
Yes, I can also confirm those diffs reported by layout-normalizer between `main` vs `vfea` (i.e. ufo2ft/compile-variable-features branch) are spurious, because hb-shape says that they do in fact match. Below an excerpt

```diff
--- Oswald-VF-main.markkern.txt 2023-12-14 12:34:44
+++ Oswald-VF-vfea.markkern.txt 2023-12-14 12:34:35
@@ -1,119 +1,33 @@
 # GPOS #

-# kern: DFLT/dflt ## 1824 PairPos rules
+# kern: DFLT/dflt ## 1572 PairPos rules
 # lookupflag LookupFlag(8)
-A -1 {-2,0} Ccedilla
 A  Q
-A -39 {-42,-34} V
-A -7 {-10,-2} v
-A -57 {-60,-52} [quotedblright,quoteright]
-Aacute -1 {-2,0} Ccedilla
 Aacute  Q
-Aacute -39 {-42,-34} V
-Aacute -7 {-10,-2} v
-Aacute -57 {-60,-52} [quotedblright,quoteright]
 Abreve  Q
-Abreve -39 {-42,-34} V
-Abreve -7 {-10,-2} v
-Abreve -57 {-60,-52} [quotedblright,quoteright]
 uni1EAE  Q
-uni1EAE -39 {-42,-34} V
-uni1EAE -7 {-10,-2} v
-uni1EAE -57 {-60,-52} [quotedblright,quoteright]
```

simoncozens commented:
Ship it! :-)

anthrotype commented on Lib/ufo2ft/featureWriters/markFeatureWriter.py:
```diff
@@ -3,13 +3,11 @@
 from ufo2ft.featureWriters import BaseFeatureWriter, ast
 from ufo2ft.util import (
     classifyGlyphs,
-    quantize,
```
@simoncozens I had to change glyphsLib's markFeatureWriter to import `quantize` from `ufo2ft.util` otherwise we get ImportError with this PR:
https://github.com/googlefonts/glyphsLib/commit/ae08b26716188d027de9c70fb9dc7c017f868dad

anthrotype replied:
oh I had forgotten that you already have a PR for that, sorry: https://github.com/googlefonts/glyphsLib/pull/957/

ok so we shall merge that one once we release this

anthrotype commented:
@simoncozens hope you don't mind if I added a `variableFeatures=True` compile option that we can use from fontmake to force disable this (https://github.com/googlefonts/ufo2ft/pull/635/commits/52ca46a4adaaf11c6cb95f0254c29f6d000a19e5), just in case
